README
== Aviary+, a Twitter Archiver

Aviary+ is a simple Ruby script to retrieve and archive your tweets. Twitter 
"claim no intellectual property rights over the material you provide to the 
Twitter service," so you should have a way of storing and repurposing your 
data.

The original Aviary script (http://github.com/mja/aviary) was written by
Mark James Adams. It only downloaded your tweets, from your timeline. I've
updated it to login via OAuth, download your mentions, your direct messages,
your sent direct messages, and any tweets that you have replied to.

== Requirements

Aviary+ requires the Launchy, nokogiri, and oauth gems.

== Running Aviary+

First, copy the config-example.yml to config.yml, then edit the default
path, for archiving tweets. The default is User/Documents/aviary, which
is probably fine for most people.

If you install your gems with rubygems, you'll need to execute Aviary+
like this:

  $ ruby -rubygems aviary.rb

If you're using Ruby 1.9, this step won't be necessary.

Next, add a Twitter account to aviary by running with
  $ ruby aviary.rb --add accountname

To add my jmartindf account I would execute it like this
  $ ruby aviary.rb --add jmartindf

Run Aviary+ with:

  $ ruby aviary.rb --updates [new|all] [--page XXX]

Aviary+ will create a directory called USERNAME and begin parsing your Twitter 
archive and downloading the raw XML representation of each tweet. The 
"--updates all" option will parse and save every tweet, while "--updates new" 
will will only request tweets that are newer than the most recent one already
present in the USERNAME direcotry.  already been downloaded. Optionally,
specify a page number to retrieve tweets starting from that page. This is
useful when restarting the script after a timeout.

Use the cat_statuses.sh script to combine all downloaded tweets for a user
into one statuses XML file.

  $ SCREEN_NAME=username ./cat_statuses.sh > username.xml
  
will read in all tweets in the username directory and concatenate them 
together in a file called username.xml. 

== Creating an events timeline

Assuming you have already put all of your status files into a single XML file
using cat_statuses.sh, you can transform it into a SIMILE timeline 
(http://simile.mit.edu/timeline/) using the assets/simili.xsl transformation 
stylesheet. View the data by modifying assets/events_template.html, replacing 
"screenname_events.xml" (Line 66) with the filename of your transformed 
statuses XML document.

This can also be done automatically using the timeline.rb script:

    $ ruby timeline.rb -u USERNAME
    
which will read in single tweets from the USERNAME's directory and construct
an events timeline (USERNAME_events.xml) in the timelines directory.
    

== What next?

Now your data is yours! Start thinking about your own way to utilize your
own data. Start thinking about your children's children's children. Will they
be accessing the possibly non-existent http://twitter.com or would they
prefer a paper record of "what you were doing"?


CHANGELOG
0.9.1
	* I'm arbitrarily starting the version numbering at 0.9.0 for my overhaul
	and 0.9.1 for this bug fix.
	* Now storing timeline tweet, mention tweets, direct messages, and sent
	direct messages in separate directories.
	* Now automatically moving xml files to the new directories, if you were
	using a previous version.
	* Before---if a tweet on your timeline was more recent than a tweet in your
	mentions, the tweet in your mentions wouldn't be downloaded (in update
	mode). The new directory structure makes it work properly.



aviary.rb
#!/usr/bin/env ruby
#
#  Created by Mark James Adams on 2007-05-14.
#  Copyright (c) 2007. All rights reserved.

require 'optparse'
require 'yaml'
require 'twitter_archiver'
require 'launchy'

AUTH_FILE = "twitauth.yml"

def prompt(default, *args)
  if not STDOUT.sync
    STDOUT.sync = true
    fix = true
  end
  print(*args)
  result = gets.strip
  if fix
    STDOUT.sync = false
  end
  return result.empty? ? default : result
end

def config(username)
  users = YAML.load_file(AUTH_FILE)
  config = Hash.new()

  key = "KYcLNiVtuTb5F55gWuVZQw"
  secret = "Vw36mGMxyhxxtfyz86UW9Fwtht4ZLmxerI4YUrRHLc"

  # Get a request token
  consumer = OAuth::Consumer.new(key, secret,
                               { :site => "http://api.twitter.com",
                                 :scheme => :header
  })
  request_token=consumer.get_request_token

  # Launch the authorization URL
  Launchy.open( request_token.authorize_url )

  # Prompt for a PIN number
  pin = prompt(nil,"Enter the Twitter PIN: ")
  
  if not pin.nil?
    # Get an access token
    access_token=request_token.get_access_token(:oauth_verifier => pin)
    config['token'] = access_token.token
    config['secret'] = access_token.secret
    users[username] = config
    File.open( AUTH_FILE, 'w' ) do |out|
      YAML.dump( users, out )
    end#File.open
  end#if not pin.nil?
end#config

begin

  CONFIG = YAML.load_file('config.yml')
  $options = {}
	$options[:tweet_path] = File.expand_path(CONFIG['archive_dir'])
  OptionParser.new do |opts|
    opts.banner = "Usage: aviary.rb --updates [new|all] --page XXX"
  
    $options[:updates] = :new
    opts.on("--updates [new|all]", [:new, :all], "Fetch only new or all updates") {|updates| $options[:updates] = updates}
    $options[:page] = 1
    opts.on("--page XXX", Integer, "Page") {|page| $options[:page] = page}
    $options[:new_account] = nil
    opts.on("--add XXX", String, "NewAccount") {|account| $options[:new_account] = account}
    $options[:only] = nil
    opts.on("--only XXX", String, "OnlyAccount") {|account| $options[:only] = account}
    $options[:debug] = nil
    opts.on("-d", "--debug", "Run verbosely") { |v| $options[:debug] = v }
  end.parse!

  if [:updates].map {|opt| $options[opt].nil?}.include?(nil)
    puts "Usage: aviary.rb --updates [new|all] --page XXX --add XXX --only XXX --debug"
    exit
  end

  if not $options[:new_account].nil?
    config($options[:new_account])
  end

  USERS = YAML.load_file(AUTH_FILE)

  case $options[:updates]
  when :new
    updates_only = true
  else :all
    updates_only = false
  end
 
  USERS.each { |user, auth|
    next unless ($options[:only].nil? or $options[:only]==user)
    token = auth['token']
    secret = auth['secret']

    ta = TwitterArchiver.new(user, token, secret, $options[:tweet_path], $options[:debug])
    ta.hark_timeline(updates_only, $options[:page]) # Get timeline
    ta.hark_mentions(updates_only, $options[:page]) # Get last 800 mentions
    ta.hark_messages(updates_only, $options[:page]) # Get direct messages sent to me
    ta.hark_messages_sent(updates_only, $options[:page]) # Get direct messages I sent
    ta.find_missing_replies # Look to see if we're missing any tweets we've replied to
    ta.try_to_download_replies # Try to download those missing messages
    ta.log_replies() # Make sure to save the list of missing messages
  }

  
rescue Errno::ENOENT => e
  puts "Whoops!"
  puts "#{e.message}" unless $options[:debug]
  puts "There is no configuration file."
  puts "Place your username and password in a file called `config.yml`. See config-example.yml."
  puts e.backtrace unless $options[:debug]
rescue RetrieveError => e
  puts "Nuts: #{e.reason}" unless not $options[:debug]
rescue StandardError => e
  puts "Caught an error... Logging replies"
  ta.log_replies()
  puts "Done"
  puts e.message unless not $options[:debug]
  puts e.backtrace unless not $options[:debug]
end





timeline.rb
require 'rubygems'
require 'tmpdir'
require 'optparse'
require 'rexml/document'
require 'fileutils'
require 'builder'

# given an array of texts, return a hash of the number of times each 
# word appears
def concordance(texts)
  count = Hash.new(0)
  texts.each do |text|
    text.split.each do |word|
      count[word] = count[word] + 1
    end
  end
  return count
end

# boil a tweet down to its least common words
def summarize(text, concordances)
  words = text.split
  text_concordances = words.map{|word| concordances[word]}
  # Include words with the least common count
  include_words_with_count = text_concordances.sort.first
  title = words.select {|word| include_words_with_count == concordances[word]}.join(' ')
  return title
end

$options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: timeline.rb -u USERNAME"
  
  opts.on("-u", "--user USERNAME", String, "Username") { |u| $options[:user] = u}
end.parse!

if [:user].map {|opt| $options[opt].nil?}.include?(nil)
  puts "Usage: timeline.rb -u USERNAME"
  exit
end

statuses_xml = File.join(Dir::tmpdir, "#{$options[:user]}.xml")

system("SCREEN_NAME=#{$options[:user]} ./cat_statuses.sh > #{statuses_xml}")

xml = File.new(statuses_xml)
doc = REXML::Document.new(xml)

texts = []

REXML::XPath.each(doc, "//text") {|element| texts << element.text}

xml.close

concordances = concordance(texts)

# make timeline data for SIMILE
FileUtils.mkdir_p('timelines')

timeline_file = File.join('timelines', "#{$options[:user]}_events.xml")

File.open(timeline_file, "w") do |timeline|
  xml = File.new(statuses_xml)
  doc = REXML::Document.new(xml)
  builder = Builder::XmlMarkup.new(:indent => 1)
  builder.instruct!
  data = builder.data do |b|
    REXML::XPath.each(doc, "statuses/status") do |status|
      text = status.elements['text'].text
      builder.event( text,
                    :start => status.elements['created_at'].text,
                    :title => summarize(text, concordances))
    end
  end
  timeline.write(data)
end






twitter_archiver.rb
#
#  Created by Mark James Adams on 2007-05-14.
#  Copyright (c) 2007. All rights reserved.

require 'nokogiri'
require 'uri'
require 'fileutils'
require 'oauth'
require 'time'

class RetrieveError < StandardError
  attr_reader :reason
  def initialize(reason)
    @reason = reason
  end
end

class RateLimit < StandardError
  attr_reader :reason
  def initialize(reason)
    @reason = reason
  end
end

class FailWhaleError < StandardError
  attr_reader :reason
  def initialize(reason)
    @reason = reason
  end
end

class TwitterArchiver
  def prepare_access_token(oauth_token, oauth_token_secret)
    key = "KYcLNiVtuTb5F55gWuVZQw"
    secret = "Vw36mGMxyhxxtfyz86UW9Fwtht4ZLmxerI4YUrRHLc"
    consumer = OAuth::Consumer.new(key, secret,
                                   { :site => "http://api.twitter.com",
                                     :scheme => :header
    })
    # now create the access token object from passed values
    token_hash = { :oauth_token => oauth_token,
      :oauth_token_secret => oauth_token_secret
    }
    access_token = OAuth::AccessToken.from_hash(consumer, token_hash )
    return access_token
  end

  def initialize(user, token, secret, root, debug=nil)
    @user = user
    @token = token
    @secret = secret
    @debug = debug

    @access_token = prepare_access_token(@token, @secret)
    @replies = Array.new()
    @lost_replies = Array.new()

    @t_prefix = File.join(root, @user,"timeline")
    @m_prefix = File.join(root, @user,"mentions")
    @dm_prefix = File.join(root, @user,"dm")
    @dm_sent_prefix = File.join(root, @user,"dm_sent")

    @api_timeline = "http://api.twitter.com/1/statuses/user_timeline"
    @api_mentions = "http://api.twitter.com/1/statuses/mentions"
    @api_dm = "http://api.twitter.com/1/direct_messages"
    @api_dm_sent = "http://api.twitter.com/1/direct_messages/sent"

    @replies_path = File.join(@t_prefix,"replies.txt")
    @lost_replies_path = File.join(@t_prefix,"lost_replies.txt")

    @t_node = "status"
    @dm_node = "direct_message"
   
    needed = cleanup_needed?

    FileUtils.mkdir_p @t_prefix
    FileUtils.mkdir_p @dm_prefix
    FileUtils.mkdir_p @m_prefix
    FileUtils.mkdir_p @dm_sent_prefix

    begin
      
      if needed
        cleanup_tweets()
      end

      File.foreach(@replies_path) { |line|
        track_reply(line.strip) unless line.nil?
      }
      File.foreach(@lost_replies_path) { |line|
        track_reply(line.strip, true)
      }
    rescue Errno::ENOENT
    end
  end

  def log_replies()
    @replies.delete("")
    @lost_replies.delete("")
    @replies.compact!
    @lost_replies.compact!
    File.open(@replies_path, 'w') { |rf|
      @replies.each { |reply|
        rf << reply.to_s << "\n"
      }
    }
    File.open(@lost_replies_path, 'w') { |rf|
      @lost_replies.each { |reply|
        rf << reply.to_s << "\n"
      }
    }
  end
 
  def track_reply(id, missing=false)
    if missing
      @replies.delete(id)
      @lost_replies.push(id)
    else
      @lost_replies.delete(id)
      @replies.push(id)
    end
    @replies.uniq!
    @lost_replies.uniq!
    log_replies()
  end

  def get_most_recent_id(dirname)
    $statuses = Array.new
    Dir.new(dirname).select {|file| file =~ /\d+.xml$/}.each{|id_xml| 
      $statuses.push(id_xml.gsub('.xml', '').to_i)
    }
    # find the most recent status
    last_id = $statuses.sort.reverse.first
    if last_id==""
      last_id=nil
    end
    last_id
  end

  def got_reply(id, missing=false)
    @replies.delete(id)
    @lost_replies.delete(id)
    @replies.uniq!
    @lost_replies.uniq!
    log_replies()
  end

  def find_missing_replies
    @replies.uniq!
    @replies.each { |id|
      pos = check_status_disk(id,@t_prefix) ### FIXME? ###
      pos2 = check_status_disk(id,@m_prefix) ### FIXME? ###
      if not pos.nil? or not pos2.nil?
        puts "Found tweet #{id}" unless @debug.nil?
        got_reply(id)
      end
    }
  end

  def hark_timeline(updates_only, page)
    hark("timeline", updates_only, page, @api_timeline)
  end

  def hark_mentions(updates_only, page)
    hark("mentions", updates_only, page, @api_mentions)

  end
  
  def hark_messages(updates_only, page)
    hark("messages", updates_only, page, @api_dm)
  end

  def hark_messages_sent(updates_only, page)
    hark("sent_messages", updates_only, page, @api_dm_sent)
  end

  def hark(type, updates_only, page, api)
    
    listening = true
    dm = false
    path = nil
    
    if api.eql?(@api_dm)
      node = @dm_node
      dm = true
      path = @dm_prefix
    elsif api.eql?(@api_dm_sent)
      node = @dm_node
      dm = true
      path = @dm_sent_prefix
    elsif api.eql?(@api_timeline)
      node = @t_node
      dm = false
      path = @t_prefix
    elsif api.eql?(@api_mentions)
      node = @t_node
      dm = false
      path = @m_prefix
    end
    
    if updates_only
      since_id = get_most_recent_id(path)
      if not since_id.nil?
        since_id_parameter = "&since_id=#{since_id}"
      else
        since_id_parameter = ""
      end
    else
      since_id_parameter = ""
    end 

    count_parameter = "count=200"
    screen_name_parameter = "&screen_name=#{@user}"
    
    while listening do # parse the account archive until we come to the last page (all)
                       # or we see a tweet we've alrady downloaded
    
      begin

        page_parameter = "&page=#{page}"
        query = count_parameter + since_id_parameter + page_parameter + screen_name_parameter
        user_timeline_url = "#{api}.xml?" + query
        
        puts "Fetching #{user_timeline_url}" unless @debug.nil?
        user_timeline_resource = @access_token.request(:get, user_timeline_url)
        user_timeline_xml = user_timeline_resource.body
        File.open(File.join(@t_prefix,"debug.log"), 'w') { |data| data << user_timeline_xml }
        puts "Retrieved #{@user} #{type} page #{page} ..."
        hp = Nokogiri(user_timeline_xml)
        tweets = (hp/node)
        if tweets.length == 0
          body = (hp/"body")
          raise FailWhaleError, "Fail Whale. Wait 5 seconds and try again" unless body.length == 0 # Fail Whale HTML page
          hash = (hp/"hash")
          raise RetrieveError, hash.at("error").inner_html unless body.length == 0
        end
        puts "Parsing #{tweets.length} tweets..."
        tweets.each {|tweet|

         id = tweet.at("id").inner_html
         
         puts "Saving tweet #{id}" unless @debug.nil?
         save_tweet_disk(id, tweet, path, dm)
        }

        unless tweets.empty?
          page = page + 1
        else
          listening = false
        end
      rescue FailWhaleError => e
        puts e.reason
        sleep(5)
        retry
      rescue RetrieveError => e
        puts "Download failed: #{e.reason}" unless @debug.nil?
        listening = false
       end
    end # listening
    
  end

  def check_status_disk(id,path)
    $statuses = Array.new
    Dir.new(path).select {|file| file =~ /\d+.xml$/}.each{|id_xml| 
      $statuses.push(id_xml.gsub('.xml', '').to_i)
    }
    return $statuses.index(id.to_i)
  end

  def get_single_tweet(id)
    tweet_url = "http://api.twitter.com/1/statuses/show/#{id}.xml"
    puts "Retrieving tweet " + id unless @debug.nil?
    tweet_resource = @access_token.request(:get, tweet_url)
    tweet_xml = tweet_resource.body
    tweet = (Nokogiri(tweet_xml)/"status").first
    error = (Nokogiri(tweet_xml)/"hash").first
    emsg = error.at('error').inner_html unless error.nil?
    if not emsg.nil? and ( emsg.include?("No status found with that ID.") or
                          emsg.include?("Not found") )
      track_reply(id, true)
    end
    if not emsg.nil? and emsg.include?("Rate limit exceeded")
      raise RateLimit, emsg
    end
    raise RetrieveError, "Tweet #{id}: #{emsg}" unless error.nil?
    return tweet
  end

  def try_to_download_replies
      @replies.each { |id|
        begin
        tweet = get_single_tweet(id)
        save_tweet_disk(id, tweet, @t_prefix, false) ### FIXME? ###
        got_reply(id)
        rescue RetrieveError => e
          puts "#{e.reason}" unless @debug.nil?
        end
      }
  end

  def save_tweet_disk(id, tweet, path, dm)
    fname = File.join(path, "#{id}.xml")
    begin
    if not check_status_disk(id,path)
      if tweet.nil?
        while tweet.nil? do
          tweet = get_single_tweet(id)
        end#while tweet.nil? do
      end#if tweet.nil?
      at = Time.parse(tweet.at("created_at").inner_html).iso8601
      puts "Saving tweet #{id}"
      File.open(fname, 'w') { |tweet_xml| tweet_xml << tweet.to_s}
      reply_to = tweet.at("in_reply_to_status_id").inner_html unless dm
      if (not reply_to.nil?) and (not reply_to.eql?(""))
        track_reply(reply_to)
      end#if reply_to
    else #status is on disk. load it and check for replies
      tweet = (Nokogiri(open(fname))/"status")
      reply_to = tweet.at("in_reply_to_status_id").inner_html unless dm
      if (not reply_to.nil?) and (not reply_to.eql?(""))
        track_reply(reply_to)
      end#if reply_to
    end#if status already exists
    rescue RetrieveError => e
      puts "Couldn't download tweet #{e.reason}" unless @debug.nil?
    end#begin block
  end#def save_tweet

  def cleanup_needed?
    t_path = File.dirname(@t_prefix)
    if File.exists?(t_path) and not File.exists?(@dm_sent_prefix)
      true
    else
      false
    end
  end

  def cleanup_tweets()
    t_path = File.dirname(@t_prefix)
    
    Dir.new(t_path).select {|file| file =~ /\d+.xml$/}.each{|id_xml| 
      full_path = File.join(t_path,id_xml)
      tweet = (Nokogiri(open(full_path))/@t_node)
      sent_by = tweet.at("user/screen_name").inner_html
      if not sent_by.eql?(@user)
        FileUtils.mv(full_path,@m_prefix)
      else
        FileUtils.mv(full_path,@t_prefix)
      end
    }
    Dir.new(@dm_prefix).select {|file| file =~ /\d+.xml$/}.each{|id_xml| 
      full_path = File.join(@dm_prefix,id_xml)
      tweet = (Nokogiri(open(full_path))/@dm_node)
      sent_by = tweet.at("sender_screen_name").inner_html
      if not sent_by.eql?(@user)
        FileUtils.mv(full_path,@dm_sent_prefix)
      end
    }
  end

end

